File: front/src/types/order.ts
/**
 * Tipos de pedido.
 *
 * O contrato novo já entrega tudo resolvido: nome do produto congelado no item
 * e nome do complemento no complemento. O Kanban não precisa mais carregar o
 * catálogo inteiro só para traduzir ids.
 */

/** Espelha `back/app/domain/enums.py::OrderStatus`. */
export type OrderStatus =
  | 'novo'
  | 'preparando'
  | 'entrega'
  | 'finalizado'
  | 'cancelado';

export type PaymentStatus =
  | 'pendente'
  | 'pago'
  | 'expirado'
  | 'cancelado'
  | 'reembolsado';

export type FulfillmentType = 'entrega' | 'retirada';

export type OrderChannel = 'whatsapp' | 'admin' | 'simulador';

/** Complemento escolhido, já com o nome congelado na venda. */
export interface OrderItemComplement {
  id?: string;
  complement_id?: string | null;
  complement_name_snapshot: string;
  extra_price_snapshot: number;
}

export interface OrderItem {
  id: string;
  product_id?: string | null;
  product_name_snapshot: string;
  unit_base_price: number;
  quantity: number;
  line_total: number;
  details?: string | null;
  complements: OrderItemComplement[];
}

/** Cobrança Pix associada ao pedido (pode não existir). */
export interface OrderPayment {
  id?: string;
  provider?: string;
  method?: string;
  amount: number;
  status: PaymentStatus;
  qr_code?: string | null;
  qr_code_base64?: string | null;
  ticket_url?: string | null;
  expires_at?: string | null;
}

export interface OrderAddress {
  rua?: string | null;
  numero?: string | null;
  bairro?: string | null;
  complemento?: string | null;
  referencia?: string | null;
}

/**
 * O pedido, em snake_case do backend ao JSX.
 *
 * Eram duas interfaces — `ApiOrder` e `Order` — com a mesma lista de ~18
 * campos em duas grafias, mais o `toOrder` reescrevendo essa lista pela
 * terceira vez. Só que `items[]` e `complements[]` atravessavam sem tradução,
 * então `product_name_snapshot` e `line_total` chegavam ao JSX em snake_case de
 * qualquer jeito: o front era misto, que é o pior dos dois mundos. `catalog.ts`
 * e `conversation.ts` já viviam em snake_case e provam que dá.
 *
 * `toOrder` continua existindo, e agora só faz o que realmente era trabalho
 * dele: converter dinheiro que chega como string, e conciliar o cliente que o
 * backend manda ora aninhado, ora achatado.
 */
export interface Order {
  id: string;
  code: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  fulfillment_type: FulfillmentType;
  channel?: OrderChannel;
  subtotal: number;
  delivery_fee: number;
  total: number;
  notes?: string | null;
  /** ISO do backend; use `formatTime`/`toMillis`, que aceitam string. */
  created_at: string;
  paid_at?: string | null;
  /** Forma aninhada que o backend às vezes manda; `toOrder` a concilia abaixo. */
  customer?: { id?: string; name?: string | null; phone?: string | null } | null;
  customer_name?: string | null;
  customer_phone?: string | null;
  address?: OrderAddress | null;
  items: OrderItem[];
  payment?: OrderPayment | null;
}

File: back/app/schemas/order.py
"""DTOs de pedido.

`OrderCreated` e `OrderSummary` são contrato com o agente de WhatsApp
(Agente 2); os demais servem ao Kanban do painel.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.domain.enums import FulfillmentType, OrderChannel, OrderStatus, PaymentStatus
from app.schemas.common import Money, ORMModel

# ---------------------------------------------------------------------------
# Contrato com o agente
# ---------------------------------------------------------------------------


class OrderCreated(BaseModel):
    """Retorno de `create_order_from_cart` — o mínimo para o agente responder."""

    id: UUID
    code: str
    subtotal: Money
    delivery_fee: Money
    total: Money
    status: OrderStatus = OrderStatus.NOVO
    payment_status: PaymentStatus = PaymentStatus.PENDENTE


class OrderSummaryItem(BaseModel):
    product_name: str
    quantity: int
    complements: list[str] = Field(default_factory=list)
    unit_price: Money
    line_total: Money
    details: str | None = None


class OrderSummary(BaseModel):
    """Resumo legível de um pedido, usado pelo agente ao responder o cliente."""

    id: UUID
    code: str
    status: OrderStatus
    payment_status: PaymentStatus
    fulfillment_type: FulfillmentType
    subtotal: Money
    delivery_fee: Money
    total: Money
    items: list[OrderSummaryItem] = Field(default_factory=list)
    customer_name: str | None = None
    customer_phone: str | None = None
    address_line: str | None = None
    notes: str | None = None
    created_at: datetime | None = None
    paid_at: datetime | None = None
    pix_qr_code: str | None = None  # copia-e-cola da última cobrança pendente


# ---------------------------------------------------------------------------
# Leitura (painel / Kanban)
# ---------------------------------------------------------------------------


class OrderItemComplementRead(ORMModel):
    id: UUID
    complement_id: UUID | None = None
    complement_name_snapshot: str
    extra_price_snapshot: Money


class OrderItemRead(ORMModel):
    id: UUID
    product_id: UUID | None = None
    product_name_snapshot: str
    unit_base_price: Money
    quantity: int
    line_total: Money
    details: str | None = None
    complements: list[OrderItemComplementRead] = Field(default_factory=list)


class PaymentRead(ORMModel):
    id: UUID
    provider: str
    provider_payment_id: str | None = None
    method: str
    amount: Money
    status: PaymentStatus
    qr_code: str | None = None
    ticket_url: str | None = None
    expires_at: datetime | None = None
    created_at: datetime | None = None


class AddressRead(ORMModel):
    id: UUID
    rua: str
    numero: str
    bairro: str
    complemento: str | None = None
    referencia: str | None = None


class CustomerRead(ORMModel):
    id: UUID
    phone: str
    name: str | None = None


class OrderRead(ORMModel):
    id: UUID
    code: str
    status: OrderStatus
    payment_status: PaymentStatus
    fulfillment_type: FulfillmentType
    channel: OrderChannel
    subtotal: Money
    delivery_fee: Money
    total: Money
    notes: str | None = None
    created_at: datetime
    updated_at: datetime | None = None
    paid_at: datetime | None = None
    cancelled_at: datetime | None = None
    customer: CustomerRead | None = None
    address: AddressRead | None = None
    items: list[OrderItemRead] = Field(default_factory=list)
    payments: list[PaymentRead] = Field(default_factory=list)


class OrderStatusUpdate(BaseModel):
    """Corpo de `PATCH /api/orders/{id}/status`."""

    status: OrderStatus

    ticket_url: str | None = None
    expires_at: datetime | None = None


__all__ = [
    "AddressRead",
    "CustomerRead",
    "OrderCreated",
    "OrderItemComplementRead",
    "OrderItemRead",
    "OrderRead",
    "OrderStatusUpdate",
    "OrderSummary",
    "OrderSummaryItem",
    "PaymentRead",
]
